syntax = "proto3";

import "google/protobuf/any.proto";

package statefun_tasks;

message Address {
    string namespace = 1;
    string type = 2;
    string id = 3;
}

message TaskRequest {
    string id = 1;
    string type = 2;
    oneof reply {
        string reply_topic = 3;
        Address reply_address = 4;
    }
    google.protobuf.Any request_state = 5;
    string content_type = 6;
    bytes data = 7;
}

message TaskResult {
    string id = 1;
    string correlation_id = 2;
    string type = 3;
    string content_type = 4;
    bytes data = 5;
    google.protobuf.Any request_state = 6;
}

message TaskException {
    string id = 1;
    string correlation_id = 2;
    string type = 3;
    string exception_type = 4;
    string exception_message = 5;
    string stacktrace = 6;
    bool retry = 7;
    statefun_tasks.TaskRequest retry_request = 8;
    google.protobuf.Any request_state = 9;
}

message TaskState {
    string type = 1;
    string content_type = 2;
    bytes data = 3;
}